### ISSUE TITLE
`Theory` syntax

### ISSUE THREAD

**Issue by digama0** [HOL-Theorem-Prover/HOL#1364]
I would like to implement a new syntax for the `new_theory` command and the prelude. This will ensure more consistent use of the header and also make it more introspectable by HOL tooling like IDEs. Proposal:

```sml
(*
  Module documentation
*)
Import fooLib barTheory
Import[open] bazTheory
Theory bla

Definition foo:
  foo = T
End
```
is equivalent to
```sml
(*
  Module documentation
*)
local open fooLib barTheory in end
open bazTheory
val _ = new_theory "bla";

Definition foo:
  foo = T
End

val _ = export_theory ();
```

Some points of interest:
* I'd like to remove the `Theory` suffix from the theory imports, since HOL itself tends to refer to theories without this suffix in e.g. qualified constant names, but I don't think it can be implemented as long as quote-filter is a preprocessor which runs without any filesystem context (unless we change the structure name itself to drop the `Theory` suffix).
* It is mandatory that there be no non-whitespace text between the beginning of the file and the `Theory` command, except for `Import` declarations. In particular, that means that all opens have to happen at the beginning of the file, and there cannot be additional commands before `new_theory`. This happens occasionally (e.g. `temp_delsimps` is often before `new_theory`), but it doesn't seem to make a difference.
* For now, holdep will continue to work as it does currently, but I would like to transition this to a warning and deprecate mid-file holdeps (e.g. using qualified identifiers for unimported files).
* The `export_theory` at the end of the file is implicit when using `Theory` syntax.

**Comment by xrchz** [comment]
I have a feeling this would make HOL exceptionally more welcoming/non-intimidating to newcomers.

**Comment by mn200** [comment]
I agree with Magnus about both the need to make the common case easy and to include support for `set_grammar_ancestry`. 

I’d also prefer to have the `Theory bla` as the first line rather than buried.  

I don’t mind dropping the `Theory` suffix though I wonder a bit about the way users will still need to remember it when they come to refer to the theorems (on the `nobind` imports). 

Finally, is it going to be easy to have a version that works for pre-bossLib files?

**Comment by digama0** [comment]
> I don’t mind dropping the `Theory` suffix though I wonder a bit about the way users will still need to remember it when they come to refer to the theorems (on the `nobind` imports).

I think we should drop the suffix on the structure name itself. But I think that can be done independently of this proposal, it's a pretty large breaking change.

**Comment by mn200** [comment]
Agreed that it can be deferred. It might at least be easy to implement (outside of Holmake) : `s/(ident)Theory/\1/g`

**Comment by mn200** [comment]
I think you are conflating two namespace levels.  This proposal is more about the SML level (the names visible in tactics for example).  The renaming you did in your example file was to do with the HOL namespaces.   Of course a more powerful syntax could do both!

**Comment by myreen** [comment]
> Finally, is it going to be easy to have a version that works for pre-bossLib files?

Could one specify that with something like:
```
Theory[bare] pair
```

**Comment by digama0** [comment]
Notes from HOL meeting:
```
Theory thy[bare]
(*
  Module documentation
*)
Ancestors
  long_name[alias LN, qualified, ignore_grammar]
  other[ignore_grammar]
  baz
  arithmetic
  list
Libs
  preamble
  Theory
  fooLib
  listSyntax

val _ = bazTheory.bla;
Definition foo_def:
  foo = T
End
```
